core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java

```diff
@@ -2407,69 +2417,125 @@ private void loadMetadataInternal(RpcContext rpcContext, LockedInodePath inodePa
-                LOG.debug(
-                    "Failed to loadMetadata because file is not in ufs:"
-                        + " inodePath={}, options={}.",
-                    descendant.getUri(), loadMetadataContext, e);
-                continue;
-              } catch (Exception e) {
-                LOG.info("Failed to loadMetadata: inodePath={}, options={}.", descendant.getUri(),
-                    loadMetadataContext, e);
-                continue;
-              }
-              if (context.getOptions().getLoadDescendantType() == LoadDescendantPType.ALL
-                  && descendant.getInode().isDirectory()) {
-                mInodeTree.setDirectChildrenLoaded(rpcContext, descendant.getInode().asDirectory());
-              }
+            if (shouldParallelize) {
+              stream = stream.parallel();
+            }
+            List<Exception> exs = stream
+                .map(child -> processLoadMetadataChildren(inodePath, rpcContext, context, child))
+                .filter(Optional::isPresent)
+                .map(Optional::get)
+                .collect(Collectors.toList());
+            if (exs.size() > 0) {
+              AggregateException e = new AggregateException(exs);
+              LOG.error("Failed to loadMetadata: inodePath={}, context={}", inodePath.getUri(),
+                  context, e);
+              throw new IOException(e);
             }
           }
           mInodeTree.setDirectChildrenLoaded(rpcContext, inodePath.getInode().asDirectory());
         }
       }
     } catch (IOException e) {
       LOG.debug("Failed to loadMetadata: inodePath={}, context={}.", inodePath.getUri(),
-              context, e);
+          context, e);
       throw e;
     }
   }
 
+  private Optional<Exception> processLoadMetadataChildren(LockedInodePath inodePath,
+      RpcContext rpcContext, LoadMetadataContext context, UfsStatus childStatus) {
+    if (PathUtils.isTemporaryFileName(childStatus.getName())) {
+      return Optional.empty();
+    }
+    AlluxioURI childURI = new AlluxioURI(
+        PathUtils.concatPath(inodePath.getUri(), childStatus.getName()));
+    if (mInodeTree.inodePathExists(childURI) && (childStatus.isFile()
+        || context.getOptions().getLoadDescendantType() != LoadDescendantPType.ALL)) {
+      // stop traversing if this is an existing file, or an existing directory without
+      // loading all descendants.
+      return Optional.empty();
+    }
+
+    try (LockedInodePath descendant = inodePath.lockDescendant(
+        inodePath.getUri().joinUnsafe(childStatus.getName()), LockPattern.READ)) {
+      LoadMetadataContext loadMetadataContext = LoadMetadataContext
+          .mergeFrom(LoadMetadataPOptions.newBuilder()
+              .setLoadDescendantType(LoadDescendantPType.NONE).setCreateAncestors(false))
+          .setUfsStatus(childStatus);
+      try {
+        loadMetadataInternal(rpcContext, descendant, loadMetadataContext, null, false);
```

Comment:
Why is parallelize `false` here? Can we leave a comment here?

Reply:
These calls set the `LoadDescendantType` as `NONE`, so setting this to true or false doesn't really have an effect. We only parallelize when the load type is `ALL`.